### ISSUE TITLE
Getting Error : invalid joystick device number

### ISSUE THREAD

**Issue by Vigneshpadoor** [pygame/pygame#2154]
I am using pygame in ubuntu 18.04 in a jetson nano .while I run my python code, I get this error 
`pygame.error: Invalid joystick device number` in the line
`controller = pygame.joystick.Joystick(0)`.
 But sometimes it works fine without giving this error. Due to this inconsistency , i  cant find the reason of the error. It would be great if someone could help

**Comment by siggisv** [comment]
Could you let us know which version of Python and pygame you are using? I think there have been some work done on the joystick code in the past months so we definitely need that info.

**Comment by Vigneshpadoor** [comment]
> Could you let us know which version of Python and pygame you are using? I think there have been some work done on the joystick code in the past months so we definitely need that info.

thank you for replying. I really appreciate you spending time to help others.
My python version is python 3.6.9 and the pygame version is 1.9.6.  And also is there any other method to read the joystick data in python other than pygame?

**Comment by ankith26** [comment]
My first suggestion would be to upgrade the pygame version to the latest Dev version like
pip3 install pygame==2.0.0dev12

**Comment by yudonglin** [comment]
Are there any possibilities that you can provide the codes? So we will be able to test the code on other platform. But like @ankith26  said, this problem may be fixed in the latest version.

**Comment by Vigneshpadoor** [comment]
Thank you guys for replying. While trying to install the new version, im getting this error

`pip3 install pygame==2.0.0.dev12`

`Defaulting to user installation because normal site-packages is not writeable
Collecting pygame==2.0.0.dev12
  Using cached pygame-2.0.0.dev12.tar.gz (5.6 MB)
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python3 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-ra44iq4x/pygame/setup.py'"'"'; __file__='"'"'/tmp/pip-install-ra44iq4x/pygame/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-7s8b5o_9
         cwd: /tmp/pip-install-ra44iq4x/pygame/
    Complete output (29 lines):`
    
    
    WARNING, No "Setup" File Exists, Running "buildconfig/config.py"
    Using UNIX configuration...
    
    /bin/sh: 1: sdl2-config: not found
    /bin/sh: 1: sdl2-config: not found
    /bin/sh: 1: sdl2-config: not found
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-ra44iq4x/pygame/setup.py", line 312, in <module>
        buildconfig.config.main(AUTO_CONFIG)
      File "/tmp/pip-install-ra44iq4x/pygame/buildconfig/config.py", line 221, in main
        deps = CFG.main(**kwds)
      File "/tmp/pip-install-ra44iq4x/pygame/buildconfig/config_unix.py", line 194, in main
        DependencyProg('SDL', 'SDL_CONFIG', 'sdl2-config', '2.0', ['sdl']),
      File "/tmp/pip-install-ra44iq4x/pygame/buildconfig/config_unix.py", line 39, in __init__
        self.ver = config[0].strip()
    IndexError: list index out of range
    
    Hunting dependencies...
    
    ---
    For help with compilation see:
        https://www.pygame.org/wiki/CompileUbuntu
    To contribute to pygame development see:
        https://www.pygame.org/contribute.html
    ---
    
    ----------------------------------------
`ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
`

Pls forgive my lack of knowledge

**Comment by Vigneshpadoor** [comment]
This is not exactly my code, but i have used the pygame something like this

    import pygame
    axis = {}
    button = {}
    hat = {}
    pygame.init()
    first = True
    while True:
        if os.path.exists("/dev/input/js0"):
            if first:
                pygame.joystick.init()
                controller = pygame.joystick.Joystick(0)
                controller.init()
                first = False
            else:
                for event in pygame.event.get():
                    if event.type == pygame.JOYAXISMOTION:
                        axis[event.axis] = round(event.value, 3)
                    elif event.type == pygame.JOYBUTTONDOWN:
                        button[event.button] = True
                    elif event.type == pygame.JOYBUTTONUP:
                        button[event.button] = False
                    elif event.type == pygame.JOYHATMOTION:
                        hat[event.hat] = event.value

            # Then i use the data as i want

Thank you in advance

**Comment by ankith26** [comment]
The issue has nothing to do with your code. 
Pip is not able to install pygame because first of all, latest Dev-12 wheels were not found for your system.
So pip decided to build pygame from source on your device, which failed because the SDL dependencies were missing.

TLDR; it’s not going to be easy to get pygame v2.0.0dev12 working on your system for now.

**Comment by ankith26** [comment]
But following the instructions at the bottom of this link
https://www.pygame.org/wiki/CompileUbuntu#pygame%20with%20sdl2%20(alpha)
Should work in your case, but it will not be easy to get this running.

**Comment by Vigneshpadoor** [comment]
I upgraded the new pygame to pygame2.0.0.dev12. But the error still exists. It mainly occurs when I disconnect and connect the joystick again

**Comment by ankith26** [comment]
Hmmm. Pygame 2 supports device hotplugging if im correct.

**Comment by ankith26** [comment]
Ahhh I got it. The issue lies in your code. You are creating a new joystick only if the variable `first` is true. But after the first joystick connects, `first` variable is set to false, and the new joystick object is not created

**Comment by ankith26** [comment]
Here is the latest docs for the joystick module
https://www.pygame.org/docs/ref/joystick.html

It will be helpful for you, especially for the new stuff added in the latest version

**Comment by Vigneshpadoor** [comment]
Thank you for your precious time guys.
The problem got solved by adding sleep for 2 seconds before and after `pygame.joystick.init()`.
Thanks a lot @ankith26 , @yudonglin , @siggisv .
